File: pipeline/extract_text.py
"""
Step 1: Extract raw text from a contract PDF.

Uses pdfplumber for text extraction.
Future extension: for scanned PDFs (no text layer), add OCR here
(e.g. pytesseract or Azure Document Intelligence).
"""

import pdfplumber


def extract_text(pdf_path: str) -> str:
    pages = []
    with pdfplumber.open(pdf_path) as pdf:
        for page in pdf.pages:
            text = page.extract_text()
            if text:
                pages.append(text)

    if not pages:
        raise ValueError(
            f"No extractable text found in: {pdf_path}\n"
            "If this is a scanned PDF, OCR support needs to be added."
        )

    return "\n\n".join(pages)

File: src/pipeline/extract_text.ts
import pdfParse from 'pdf-parse'
import { readFile } from 'fs/promises'

export async function extractText(pdfPath: string): Promise<string> {
  const buffer = await readFile(pdfPath)
  const data = await pdfParse(buffer)

  const text = data.text.trim()
  if (!text) {
    throw new Error(
      `No extractable text found in: ${pdfPath}\n` +
      'If this is a scanned PDF, OCR support needs to be added.'
    )
  }

  return text
}
